### PR TITLE
Get rid of TSParenthesizedType nodes before proceeding with formatting

### LINKED ISSUE #3831: TypeScript: Unstable case with ParenthesizedType UnionType
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACGATwzjwDMII8BePAQW2wENCAeACgB0o88AfPMG69eAfVFESUJgFs4SPJxAAhJgC8lAGmF4AvjoFCeI8ZISz5ilQFc1GkNuP6oASgB83ByAgYYAS2gAZ2RQJkYIAHcABTCEYJQmADcIPwATLyZAmGQyJgAbQLhNEAAjZjAAazgYAGUMJjA-KABzZBhsayKQVIgwHPzC4qbC3CjmZpkmfoKugCtAtGVyqtqLABkmuGnB72sYDD2AJm2u+uwR5BAzQLBsP18vDDvYAHU0mAALZAAOAAZip4QQovZgYS5POAjRJbYrYOAAR2sfjhYyYEymSFyM2KhRkfjaHS6gSazTycAAitYIPATsUYEwSm9Up9kIc6cw-HkSQBhCAySaXKDQGEgayFAAqDPiWMKul0QA)
```sh
--parser typescript
```

**Input:**
```tsx
export type foo = Array<(
  | {
    __typename: "Baz",
  }
  | {
    __typename: "Buzz",
  }
)>

```

**Output:**
```tsx
export type foo = Array<

    | {
        __typename: "Baz";
      }
    | {
        __typename: "Buzz";
      }
>;

```

**Expected behavior:**
If you run the result through prettier again, you get the stable result of:
```tsx
export type foo = Array<
  | {
      __typename: "Baz";
    }
  | {
      __typename: "Buzz";
    }
>;
```

[ASTExplorer](https://astexplorer.net/#/gist/f53bd5cfed45137b5e495daa13379746/6666eca993bb8479142af1b67baeeb2c4bb4b2e5)

### LINKED ISSUE #3941: Doesn't remove unnecessary parens around single elements in some typescript type expressions
**Prettier 1.10.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACeBnGbAXmwAooBXAWwCM4AnASmwB8yD6BLKAc2bdIBDKBBgALBowDcAHViYc+QiXLU6TbADJ2MLr2bahI8ZKkgANCAgYYnaHmShB9ehADuABWcIHKQQDcITgATCxBBAmQAM0EAGzw4Sxp6QTAAazgYAGUMVO4eZF0KRJBgiDBouITLbgT6GA8UnipBSviSgCs8AA8AIRT0zKzBKjgAGW44NuqrChgMOYAmaZLc+jrkVAU8MC4bMIw9GAB1EPFkAA4ABktDiATjlIxNw7g6-ynLejgARwpOb6NQTNVpIGLtSwJKicQr0YqQ-KxOAARQooimYKqJRgghop2C5yQi0sukEnFi+QAwhAqC1NiIoJ8QBQEgAVXG+cEJAC+3KAA)
```sh
--parser typescript
```

Narrows the issue for #3939 

**Input:**
```tsx
type test = (number) | (string) | (another);
type test = (number) & (string) & (another);
```

**Output:**
```tsx
type test = (number) | (string) | (another);
type test = (number) & (string) & (another);

```

**Expected behavior:**
```tsx
type test = number | string | another;
type test = number & string & another;

```

Also, works correctly for flow https://github.com/prettier/prettier/issues/3939#issuecomment-364448467

### PR DESCRIPTION
**Pull request by thorn0**
Something I was talking about [here](https://github.com/prettier/prettier/pull/6381#issuecomment-534175434).

See also https://github.com/babel/babel/issues/9546

Closes #3831
Closes #3941

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://deploy-preview-6605--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEMCeAHOACASnARwFc4BnGAFUxwF5sB6AKgEZH6AfJgJjewHIA4gFEKfbO34AJIQEEAInwDcAHVjVsAWUIlyVLNjpNWHbr0Eilq1en0AxA9gDaq7OOwyPnr958yXbgCEg4JDQsID-CQBhGNi4+ISoyOw5VLT0jMy5ABpk33zfZPDi8NUAXVUQbJAIDBgAS2hSZFAAQwAndogAdwAFDoRmlFaANwh6gBMqkAAjdtawAGs4GABlDAX6qABzZBh2kmqACxgAWwAbAHUj+vhSDbA4VcHb+pHbtGRwUmbqrdI4O0YL15ttTq1kAAzVrnAHVABWpAAHgF5ksVqtWqc4AAZLZwKEwuEgRFI1ZbbbnOAARSIEHghNhcGqG3aAPaXxsZDA7XqdWmGF5sEukxgR2QAA4AAwsroAy7zDBfQVkQEjAnVYj0uAg2pDECtUgAWigcDgE3N03a2nq1pBrTBEKQ0KZ1QBp3qewOzJApApVNp2sZxJgrRmIomYuQXGq+1a9XOFKiEFO4K+UGgGpARABFDDQxdxJGJAAklALbBVjy+TAZOXVugqcG4ABfFtAA)✨**

### CODE REVIEW

**Comment by evilebottnawi** [comment]
Ready for review?

**Comment by thorn0** [comment]
nope, it's a draft

**Comment by lydell** [comment]
This sounds like a very good idea. :+1:

**Comment by lydell** [comment]
This sounds like a very good idea. :+1:

**Comment by thorn0** [comment]
I'd say we should merge #6381 and #6604 first. If any issues arise with them, I'll fix them here, in this PR. That way we'll keep changelog entries and tests from those two PRs, and this PR will be just a refactoring, not a "user-facing change".

**Comment by thorn0** [comment]
I wonder if I should remove [these parts](https://codecov.io/gh/prettier/prettier/compare/385c2dcc4fb7ee92f36abfa8db4eb31e1a65a7c3...9659bc2e63245c37a20a2d72f4c2e2848cab3700/src/src/language-js/clean.js) from `clean.js`.

File: CHANGELOG.unreleased.md [hunk 1]
```diff
@@ -813,13 +852,13 @@ Previously, the flag was not applied on html attributes.
...
-#### TypeScript: Fix incorrectly removes double parentheses around types ([#6604] by [@sosukesuzuki])
+#### TypeScript: Fix incorrectly removed double parentheses around types ([#6604] by [@sosukesuzuki])
 
 <!-- prettier-ignore -->
 ```ts
 // Input
 type A = 0 extends ((1 extends 2  ? 3 : 4)) ? 5 : 6;
-type B = ((0 extends 1 ? 2 : 3)) extends 4 ? 5 : 6:
+type B = ((0 extends 1 ? 2 : 3)) extends 4 ? 5 : 6;
 type C = ((number | string))["toString"];
```
Comment (sosukesuzuki) [inline CHANGELOG.unreleased.md]: Thank you!:smile:

**Comment by evilebottnawi** [comment]
Great job!

**Pull request merged**